### ISSUE TITLE
Redesign of the configuration system

### ISSUE THREAD

**Issue by yujinakayama** [bbatsov/rubocop#1151]
As we know the current configuration system is flawed, I'm planning a redesign of the system. Currently this is a draft and feedbacks are welcomed.

### Not project-aware

The current system allows you to customize configuration in a generic directory-based way. Practically `.rubocop.yml` is put in a project root directory, but technically it doesn't indicate where the project root is. Since generally configuration is a project-specific thing and should not be overridden by another configuration outside of the project, the current system's versatility unnecessarily makes things complex:

* Complex handling of `Exclude`
  * We cannot stop searching `.rubocop.yml` at the project root, and need to check outside of the project for this.
  * Nested projects (e.g. another project using RuboCop in `vendor`)
* Handling of options that affect to the entire project (e.g. `RunRailsCops`)
  * What if it's specified in a `.rubocop.yml` in a subdirectory?

### Complex inheritance system

The inheritance provides a sort of abstraction, but it seems to be overkill for a project-specific thing.

* Complex hash merge logic
* It's obscure.
  * Sometimes you may think “Why isn't this code pointed out as an offense? Oh, there's another `.rubocop.yml` in the `spec` directory!”
* `Include/Exclude` paths and the config file paths
  * It's confusing whether the paths are handled relative to the inherited config or the inheriting config.

### Not programmable

Since it's YAML:

* We need to implement every feature for new demands.
* Users need to learn the RuboCop-specific configuration syntax.
  * Even if the feature can be done with pure Ruby.
* There's no way to iterate cops.
  * Some people might want a way to disable all cops once and enable only a few cops in an opt-in approach.
* Array data (e.g. `Include/Exclude`) in the default configuration are reset by user's data.
  * You cannot _append_ elements to the default array.
* Ugly regexp `!ruby/regexp /^some_pattern$/`

## New design

The new design needs to solve all the problems above.

* It's Ruby
* Only one file for a project
  * Put in a project root directory
  * Searched upward from the current working directory, so that nested projects can be simply handled.
  * `Rubocopfile` (though this may spread the misspelled `Rubocop` :P) or `.rubocop.rb`?
* Base configuration can be partially overridden with block for specific paths.
* Should be less magical while keeping the readability as much as possible.

Though there's a tradeoff between readability and magical syntax (and it's the most difficult decision in this design) and currently I'm not sure about it, here's an example:

```ruby
# coding: utf-8

# The normal `main` context here.

# Plugins and baseline configs can be loaded
# with the pure Kernel#require.
require 'rubocop/some_plugin'
require 'my_organization_baseline_rubocop_config'

# This would be `RuboCop.configure(:todo)` in the rubocop-todo file
# so that we can ignore the configuration inside of the block
# when regenerating it.
RuboCop.configure do |config|
  # This is an instance_eval'ed context.
  # Actually I didn't want to do so and wanted to make this less magical,
  # but possible alternative ways are ugly:
  #   config['Style/LineLength'].enabled = false
  #   config::Style::LineLength.enabled = false
  #   config.Style.LineLength.enabled = false

  # You can append elements to the array
  # while keeping the default elements.
  config.includes << 'Guardfile'

  # Of course they can be reset if you want.
  config.excludes = []

  # Disable all style cops.
  Style.enabled = false
  # This is equivalent to:
  Style.all { |cop| cop.enabled = false }

  # These constants are not actual cop classes,
  # probably something like RuboCop::Config::Style::LineLength.
  Style::LineLength.tap do |cop|
    cop.enabled = true
    cop.max = 100
  end

  Lint::AssignmentInCondition.allow_safe_assignment = false

  # Override some options in files matching the pattern.
  config.override(/^spec\//) do
    Lint::Void.enabled = false
  end

  config.override('some/special/file.rb') do
    # Disable all cops.
    Cop.enabled = false
  end
end
```

What do you think? @bbatsov @jonas054

**Comment by sch1zo** [comment]
I there any update on this topic? Or anybody really working on it?
Anyway I looked into how rubocop handels its config internally and played around with an idea on how I would handle cop configuration and use a ruby configuration file.

The result of this can be found in this gist: https://gist.github.com/sch1zo/cd06dd8b0bd641cdfe31

Some points to note:
- I would not add support for multiple config files. by which I mean the config has to be searched/merged and applied for each folder
- one config file called `.rubocop.rb` in the project root, which is search upwards till a git/hg root is found(or we reach the home directory / root)
- global defaults either in Configuration class or as an internally required second default config (although without cop configuraton) 
- a Cop knows its own config(stored as class instance variables)
  - including its defaults
  - and allowed options (eg. for enforced_style) -> config validation/sanity checks
  - this way if I want to know how a cop works I just look at the cop and can see right there what its defaults are, what I can configure and what the allowed options are
- no duplication of Cop classes needed just to make configuration work -> there is basically no extra code after the cops know their own config
- not included in this PoC but commandline params should be folded into the `Configuration` instance so the rest of the code only has to deal with one thing
- I left out the override functionally from the original proposal for now but I like the idea behind it
- the rubocop cli could provide help in debugging config issues (eg. show enabled cops for a file or config of a cop)

Let me know what you think about my PoC. I'll probably play around with it some more anyway and look into how much stuff in rubocop would need to change to introduce something like this.

**Comment by mvz** [comment]
I'd like to add another problem with this proposal:

A simple configuration file in YAML or some other generally parsable format allows external tools to read the file without having to worry about security problems. For example, there is at least one code reviewing service (pullreview.com) that uses RuboCop to analyse pull requests. They are currently not using the project's RuboCop configuration but instead have their own. With configuration in YAML format, they would at some point be able to safely merge the project's configuration with their own. Configuration as Ruby would make this, or even running RuboCop with the project's configuration much more dangerous.

If people want to make programmatic rules for their RuboCop configuration, I think they should create a separate rake task to generate the .rubocop.yml file.

**Comment by bbatsov** [comment]
> For example, there is at least one code reviewing service (pullreview.com) that uses RuboCop to analyse pull requests.

HoundCI uses RuboCop as well. I see CodeClimate are planning to add style check to their platform and I'd be surprised if they don't use RuboCop as well (as most of their checks are done with other open-source tools).

Having two config systems would introduce significant overhead IMO. To be honest I haven't thought about this issue in a while (super busy at work, etc), but we'll definitely keep in mind the external services relying on the existing format.

**Comment by bbatsov** [comment]
Without @yujinakayama it seems this is going nowhere and I'll just close it.
